From b3a33c235fd2267fe9c48cc9fab5e7b642cdc85d Mon Sep 17 00:00:00 2001
From: Jeevaka Badrappan <jeevaka.badrappan@intel.com>
Date: Tue, 24 Sep 2013 11:25:34 +0300
Subject: [PORT FROM MAIN] Telephony: Fix PDP context not deactivated on corner cases

BZ: 108902

If the user disables "Data enabled" in out of service
state, data deactivation request is ignored due to
device out of service. When the device gets back to
registered state, browsing is allowed even when the
data is disabled by user.

Check the data enabled setting and cleanup all connections
if data is disabled. Attached state is notified as this
may be needed to update UI.

Change-Id: I0fcef01fcd31375d700103e6a6e9a4cbff5818ec
Orig-Change-Id: I4ee979029eaa5eec97a218866cb56f8638f1cf9f
Category: aosp improvement
Domain: Telephony-Telephony-FWK
Origin: internal
Upstream-Candidate: yes
Signed-off-by: Jeevaka Badrappan <jeevaka.badrappan@intel.com>
Signed-off-by: Arun Ravindran <arun.ravindran@intel.com>
---
 .../telephony/dataconnection/DcTracker.java        |   15 +++++++++++++++
 1 files changed, 15 insertions(+), 0 deletions(-)

diff --git a/src/java/com/android/internal/telephony/dataconnection/DcTracker.java b/src/java/com/android/internal/telephony/dataconnection/DcTracker.java
index d001eff..48b3107 100644
--- a/src/java/com/android/internal/telephony/dataconnection/DcTracker.java
+++ b/src/java/com/android/internal/telephony/dataconnection/DcTracker.java
@@ -542,6 +542,21 @@ public final class DcTracker extends DcTrackerBase {
         if (DBG) log("onDataConnectionAttached");
         mAttached.set(true);
         if (getOverallState() == DctConstants.State.CONNECTED) {
+            if (!getAnyDataEnabled()) {
+                /*
+                 * If the user disables "Data enabled" in out of service state,
+                 * data deactivation request is ignored due to device out of
+                 * service. When the device gets back to registered state,
+                 * browsing is allowed even when the data is disabled by user.
+                 * Check the data enabled state and cleanup all connection if data
+                 * is disabled. ATTACHED state is notified as this may be needed
+                 * to update UI.
+                 */
+                onCleanUpAllConnections(Phone.REASON_DATA_DISABLED);
+                notifyDataConnection(Phone.REASON_DATA_ATTACHED);
+                return;
+            }
+
             if (DBG) log("onDataConnectionAttached: start polling notify attached");
             startNetStatPoll();
             startDataStallAlarm(DATA_STALL_NOT_SUSPECTED);
-- 
1.7.4.1

